rust/lance/src/dataset/write/merge_insert/exec/indexed_lookup.rs

```diff
@@ -0,0 +1,470 @@
+        with_row_addr: bool,
+    ) -> DFResult<Arc<dyn ExecutionPlan>> {
+        // Step 1: MapIndexExec to get row IDs from the index
+        let mut index_mapper: Arc<dyn ExecutionPlan> = Arc::new(MapIndexExec::new(
+            dataset.clone(),
+            key_column.to_string(),
+            index.name,
+            key_projection,
+        ));
+
+        // Step 2: Add row address if requested
+        if with_row_addr {
+            let pos = index_mapper.schema().fields().len();
+            index_mapper = Arc::new(AddRowAddrExec::try_new(index_mapper, dataset.clone(), pos)?);
+        }
+
+        // Step 3: TakeExec to fetch target data for matched rows
+        let projection_schema = dataset
+            .schema()
+            .project(
+                &projection_columns
+                    .iter()
+                    .map(|s| s.as_str())
+                    .collect::<Vec<_>>(),
+            )
+            .map_err(|e| datafusion::error::DataFusionError::External(Box::new(e)))?;
+        let projection = dataset.empty_projection().union_schema(&projection_schema);
+
+        let take_result = TakeExec::try_new(dataset.clone(), index_mapper, projection)?;
+        let target_exec: Arc<dyn ExecutionPlan> = match take_result {
+            Some(take) => Arc::new(take),
+            None => {
+                return Err(datafusion::error::DataFusionError::Plan(
+                    "TakeExec returned None - no columns to take".to_string(),
+                ));
+            }
+        };
+
+        // Step 4: Project to output schema order [projection_columns, _rowid, _rowaddr]
+        let target_schema: Schema = dataset.schema().into();
+        let mut target_columns: Vec<Arc<Field>> = Vec::new();
+        for col_name in projection_columns {
+            let field = target_schema.field_with_name(col_name).map_err(|_| {
+                datafusion::error::DataFusionError::Plan(format!(
+                    "Projection column '{}' not found in target schema",
+                    col_name
+                ))
+            })?;
+            target_columns.push(Arc::new(field.clone()));
+        }
+        target_columns.push(Arc::new(ROW_ID_FIELD.clone()));
+        if with_row_addr {
+            target_columns.push(Arc::new(ROW_ADDR_FIELD.clone()));
+        }
+
+        Ok(Arc::new(project(
+            target_exec,
+            &Schema::new(target_columns),
+        )?))
+    }
+
+    /// Build the complete join pipeline with an already-created ReplayExec.
+    ///
+    /// Pipeline structure:
+    /// ```text
+    /// HashJoinExec (Left or Inner join, CollectLeft mode)
+    ///   source (ReplayExec) <- LEFT, collected into hash table
+    ///   target (lookup result) <- RIGHT, streams and probes
+    /// ```
+    ///
+    /// The source_replay must be a ReplayExec created externally. This ensures
+    /// the same Arc is used for both the exposed child and internal DAG.
+    fn build_join_pipeline_with_replay(
```

Comment:
for fully indexed case, I ended up doing the join within this exec node so that it can more easily pass around the replay node.